:memo: matcher documentation (#305)
alakarteio/k-ramel: packages/k-ramel/API.md

| hideRedux | `boolean` | `true` | actions and selectors from [`k-redux-factory`](https://github.com/alakarteio/k-redux-factory) are used without specifying `dispatch` or `getState`. |
| enhancer | redux enhancer | `undefined` | usual `compose` and `applyMiddlare` you already use with Redux can be injected here (like router, redux-saga, etc). <br />`compose` and `applyMiddleware` from Redux are exposed by the lib. To use them:<br /> ```import { compose, applyMiddleware } from 'k-ramel'```. |
| init | `object` | `undefined` | the default value of your store. |

### types
```js
import { types } from 'k-ramel'
```

> You can use when of our types as reducers (+ selectors and reactions).

Reviewer: I don't understand

Author: one of

Reviewer: that's what I was thinking about ^^

> You can use when of our types as reducers (+ selectors and reactions).

Author: ```suggestion
You can use one of our types as reducers (+ selectors and reactions).
```

A `matcher` can be one of these types:
 - a `string`: in which case the action.type will be tested against your given string
    - ex: `when('@@krml/INIT')` will react to action `{ type: '@@krml/INIT', payload: 'whatever' }`
 - a `regexp`: in which case the action.type will be tested against your given regexp
    - ex: `when(/@@krml.*/)` will react to action `{ type: '@@krml/WHATEVER', payload: 'whatever' }`
 - a `function(action, store)`: in which case the function should return a truthy value to trigger the reaction.

A `reaction` must follow this signature: `(action, store, drivers) `

#### Exemple 1: with a string
In this example we react to the k-ramel initialisation to load the **Adventure time** show description from **TVMAZE**.

```js
when('@@krml/INIT')((action, store, drivers) => {
  drivers.http('TVMAZE').get('http://api.tvmaze.com/shows/290')
})
```

#### Exemple 2: with a regexp
In this example we react to the http driver response to retrieve **Adventure time** show description that the API responds.
We take care of looking for action.status so we make sure the request is ok!

```js
when(/@@http\/.*TVMAZE.*/)((action, store) => {

>   // if the request is failed, we are doing nothing

Reviewer: ```suggestion
  // if the request fails, we do nothing
```

  if (action.status < 200 || action.status >= 400) return

>   // in others cases we register our show description to the store

Reviewer: ```suggestion
  // in other cases we register our show description to the store
```

  store.show.set(action.payload)
})
```

#### Example 3: with a function
We are doing the same example given previously but using a function matcher.

>  - The first matcher used is a regexp (like previous exemple): we are making sure to process a HTTP action.

Reviewer: ```suggestion
 - The first matcher used is a regexp (like previous exemple): we ensure to process a HTTP action.
```

>  - The second matcher is a function to make sure the request is not failed!

Reviewer: ```suggestion
 - The second matcher is a function to ensure the request has not failed!
```

    * Then, we don't need to check this in our reaction

```js
when(
  /@@http\/.*TVMAZE.*/,
  (action) => (action.status >= 200 && action.status < 400),
)((action, store) => {
  // we register our show description to the store
  store.show.set(action.payload)
})
```

#### Example 4: with a function, use the store
In the function matcher you have access to the store.
It can be useful when you want to match a specific Redux action since our actions types are published. (if you use `types`).

```js

> when((action, store) => action.type === store.show.SET)(() => { console.log('show is set!') })

Reviewer 2: It's cool to see that it's already possible to do it :-)

I will try to use k-ramel for a while in a project and I will make a proposition if I find it useful to add a custom matcher for a better readability.

Author: thank you

Author: If we add an helper, we can add it in a minor version
